### PR TITLE
General improvement part3: separate NN structure for graph and node level output

### PR DESCRIPTION
**Pull request by pzhanggit**
1. Separate NN structures for graph and node level predictions
2. Make NN structures for graph and node configurable
3. Tested for node level prediction locally, though not in CI test yet

### CODE REVIEW

File: test_trained_model.py [hunk 1]
```diff
@@ -41,6 +41,9 @@ def test_trained_model(config_file: str = None, chosen_model: torch.nn.Module =
         else:
             raise ValueError("Unknown output type", output_type[item])
         config["NeuralNetwork"]["Architecture"]["output_dim"].append(dim_item)
+    config["NeuralNetwork"]["Architecture"]["output_type"] = config["NeuralNetwork"][
```
Comment (streeve) [inline test_trained_model.py:44]: Why do we need to add to the config?

File: models/GATStack.py [hunk 2]
```diff
@@ -12,20 +12,22 @@ def __init__(
         self,
         input_dim: int,
         output_dim: list,
+        output_type: list,
         num_nodes: int,
-        hidden_dim: int = 16,
```
Comment (streeve) [inline models/GATStack.py:16]: I assume this is for consistency across models

File: examples/configuration.json [hunk 3]
```diff
@@ -21,7 +21,19 @@
             "max_neighbours": 5,
             "hidden_dim": 5,
             "num_conv_layers": 6,
-            "num_sharedlayers": 2,
+            "heads": {
```
Comment (streeve) [inline examples/configuration.json:24]: I think I need you to explain this to me in more detail

**Review by streeve — COMMENTED** [review]
Can you post accuracy details from local runs? I'm curious how much this improves results. I also think I need a better understanding before I can approve

File: test_trained_model.py (hunk 1 above)
Comment (pzhanggit) [reply test_trained_model.py:44]: Because different NN structures are designed for graph and node outputs.

**Comment by pzhanggit** [comment]
@streeve See f6c97c3 for new changes we discussed earlier.

**Review by streeve — COMMENTED** [review]
Looks great - I'll approve if you add a comment that you tested with node-level outputs (not in the CI yet) and fix the two really minor things below

File: models/Base.py [hunk 4]
```diff
@@ -76,10 +77,12 @@ def _multihead(
 
     def node_features_reshape(self, x, batch):
         """reshape x from [batch_size*num_nodes, num_features] to [batch_size, num_features, num_nodes]"""
-        batch_size = batch.max() + 1
         num_features = x.shape[1]
+        self.batch_size = batch.max() + 1
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply models/Base.py:81]: Yeah, agreed, I will change it.  I think neither of them should be a member variable unless it shows big improvement in performance.

File: examples/ci.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
     "Dataset": {
         "name": "unit_test",
         "format": "unit_test",
-        "num_nodes": 10,
+        "num_nodes": 8,
```
[the remark this answers is not in the document]
Comment (pzhanggit) [reply examples/ci.json:5]: I would keep it. It comes together with dataset.

**Comment by pzhanggit** [comment]
> Looks great - I'll approve if you add a comment that you tested with node-level outputs (not in the CI yet) and fix the two really minor things below

See f9a5b34 and also the update description.

File: examples/ci.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (streeve) [reply examples/ci.json:5]: I didn't realize this has no effect on the unit test - problem for another PR

**Review by streeve — APPROVED** [review]
